CnCNet/Net/PeerToPeer/PeerToPeerUtil.cs

```diff
@@ -1,135 +1,121 @@
+    private readonly ILogger logger;
+
+    public PeerToPeerUtil(ILogger<PeerToPeerUtil> logger)
     {
-        const int COUNTER_RESET_INTERVAL = 60 * 1000; // Reset counter every X ms
-        const int MAX_REQUESTS_PER_IP = 20; // Max requests during one COUNTER_RESET_INTERVAL period
-        const int MAX_CONNECTIONS_GLOBAL = 5000; // Max amount of different ips sending requests during one COUNTER_RESET_INTERVAL period
-        const int STUN_ID = 26262;
+        this.logger = logger;
+        new Random().NextBytes(sendBuffer);
+        Array.Copy(BitConverter.GetBytes(IPAddress.HostToNetworkOrder((short)StunId)), 0, sendBuffer, 6, 2);
+    }
 
-        int ListenPort;
-        UdpClient Client;
-        Task ReceiveTask;
-        Dictionary<int, int> ConnectionCounter = new Dictionary<int, int>(MAX_CONNECTIONS_GLOBAL);
-        Timer ConnectionCounterTimer = new Timer(COUNTER_RESET_INTERVAL);
-        object _lock = new object();
-        byte[] SendBuffer = new byte[40];
+    public Task StartAsync(int listenPort, CancellationToken cancellationToken)
+    {
+        connectionCounterTimer.Elapsed += (_, _) => _ = ResetConnectionCounterAsync(cancellationToken);
+        connectionCounterTimer.Enabled = true;
 
+        return StartReceiverAsync(listenPort, cancellationToken);
+    }
 
-        public static Task StartNew(int listenPort) { return new PeerToPeerUtil(listenPort).Start(); }
+    public void Dispose()
+    {
+        connectionCounterSemaphoreSlim.Dispose();
+        connectionCounterTimer.Dispose();
+    }
 
-        public PeerToPeerUtil(int listenPort)
+    private async Task ResetConnectionCounterAsync(CancellationToken cancellationToken)
+    {
+        try
         {
-            ListenPort = listenPort;
-            new Random().NextBytes(SendBuffer);
-
-            Array.Copy(
-                BitConverter.GetBytes(IPAddress.HostToNetworkOrder((short)STUN_ID)),
-                0,
-                SendBuffer,
-                6,
-                2);
-
-            ConnectionCounterTimer.Elapsed += new ElapsedEventHandler((sender, e) => ResetCounter());
-            ReceiveTask = new Task(() => SyncReceive(), TaskCreationOptions.LongRunning);
-        }
+            if (cancellationToken.IsCancellationRequested)
+                connectionCounterTimer.Enabled = false;
 
-        public Task Start()
-        {
-            if (!ReceiveTask.Status.HasFlag(TaskStatus.Running))
+            await connectionCounterSemaphoreSlim.WaitAsync(cancellationToken).ConfigureAwait(false);
+
+            try
             {
-                ReceiveTask.Start();
-                ConnectionCounterTimer.Enabled = true;
+                connectionCounter.Clear();
             }
-
-            return ReceiveTask;
-        }
-
-        private void ResetCounter()
-        {
-            lock (_lock)
+            finally
             {
-                ConnectionCounter.Clear();
+                _ = connectionCounterSemaphoreSlim.Release();
```

Comment:
Also contrary to MS-encouraged practice of using discards like this we don't actually use discards, we think it makes the code less readable, so please amend them here and all over the code.

Reply:
I'm on MS's side with this but I'll adapt it.

[Remove unnecessary expression value (IDE0058)](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0058):
> This improves code clarity by explicitly showing the intent to discard an unused value.